File: library/src/main/java/io/github/pelmenstar1/rangecalendar/SavedState.java
package io.github.pelmenstar1.rangecalendar;

import android.os.Parcel;
import android.os.Parcelable;
import android.view.AbsSavedState;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

final class SavedState extends AbsSavedState {
    public int ym;

    public int selectionType;
    public int selectionYm;
    public int selectionData;

    public SavedState(@NotNull Parcelable superState) {
        super(superState);
    }

    public SavedState(@NotNull Parcel source) {
        super(source);

        ym = source.readInt();
        selectionType = source.readInt();
        selectionYm = source.readInt();
        selectionData = source.readInt();
    }

    @Override
    public void writeToParcel(@NotNull Parcel dest, int flags) {
        super.writeToParcel(dest, flags);

        dest.writeInt(ym);
        dest.writeInt(selectionType);
        dest.writeInt(selectionYm);
        dest.writeInt(selectionYm);
    }

    @NotNull
    public static final Parcelable.Creator<SavedState> CREATOR = new Creator<SavedState>() {
        @Override
        @NotNull
        public SavedState createFromParcel(@NotNull Parcel source) {
            return new SavedState(source);
        }

        @Override
        public @Nullable SavedState @NotNull [] newArray(int size) {
            return new SavedState[size];
        }
    };
}

File: library/src/main/java/io/github/pelmenstar1/rangecalendar/SavedState.kt
package io.github.pelmenstar1.rangecalendar

import android.os.Parcelable
import android.os.Parcel
import android.os.Parcelable.Creator
import android.view.AbsSavedState

internal class SavedState : AbsSavedState {
    var ym = YearMonth(0)
    var selectionType = 0
    var selectionYm = YearMonth(0)
    var selectionData = 0

    constructor(superState: Parcelable) : super(superState)

    constructor(source: Parcel) : super(source) {
        source.run {
            ym = YearMonth(readInt())
            selectionType = readInt()
            selectionYm = YearMonth(readInt())
            selectionData = readInt()
        }

    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        super.writeToParcel(dest, flags)

        dest.run {
            writeInt(ym.totalMonths)
            writeInt(selectionType)
            writeInt(selectionYm.totalMonths)
            writeInt(selectionData)
        }
    }

    companion object {
        @JvmField
        val CREATOR: Creator<SavedState> = object : Creator<SavedState> {
            override fun createFromParcel(source: Parcel) = SavedState(source)
            override fun newArray(size: Int) = arrayOfNulls<SavedState>(size)
        }
    }
}